[Feature] Add unlock token mechanism

CaptainVincent opened the pull request:
**Unlock token mechanism is base on `Token Economics` in [Oasis ETH gas token](https://hackmd.io/@PIv_zaEuR2e5PtCzEc4HRA/HkbOgO47v)**

- Summary
There is a total fixed supply of 21 million OETHs. The blockchain unlocks 525,000 OETHs every month in the first 20 months after the mainnet launch. The monthly release is cut to 1/2 every 20 months. It takes perpetuity to release the entire 21 million OETHs.

- Implementation (update 2020.11.6)
  1. ~~We expect unlock token once every month when runtime get the first tx in the month.~~
  2. Use fixed 30 days as unit round. The release amount will cut to 1/2 every 20 rounds.
  3. Release amount in a single round is depending on account for what is the percentage in this round.
  4. ~~If runtime doesn't get any tx in expecting month,~~ The unhandled release quota will pending until getting any transaction. 
  5. We keep some information (initial timestamp, ~~unlock round, unlock token amount~~, pending round, unlocked ticks) in special account's storage for making every compute nodes reach a consensus.
  > - `initial timestamp` means the start time to apply unlock token mechanism.
  > - `pending round` means which round is not yet fully unlocked.
  > - `unlocked ticks` means already unlocked ticks/seconds in last unlocked round. 
  > - Below is more detail of the used terms in the implementation about calculating to unlock amount.
![fig](https://user-images.githubusercontent.com/12159865/99381461-260f1f80-2906-11eb-9938-644a0addab45.jpg)

- Testing
run `cargo test --features="test"`.
The result should be the same as the following content. 
```
(omitted)

running 8 tests
test fund::tests::test_try_unlock_1_round ... ok
test fund::tests::test_try_unlock_10_rounds ... ok
test fund::tests::test_try_unlock_sequential_3_ticks ... ok
test fund::tests::test_try_unlock_zero ... ok
test fund::tests::test_try_unlock_random_txs_20rounds ... ok
test fund::tests::test_try_unlock_random_txs_cross_20rounds ... ok
test fund::tests::test_try_unlock_100_rounds ... ok
test fund::tests::test_try_unlock_1000_rounds ... ok

test result: ok. 8 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

(omitted)
```

CaptainVincent committed:
[Feature] Add unlock token mechanism

hydai commented on src/fund.rs:
```diff
@@ -0,0 +1,63 @@
+use std::convert::TryInto;
+use byteorder::{BigEndian, ReadBytesExt};
+
+fn timestamp_to_mt(timestamp:i64) -> MonthTuple{
```
Should we run a fmt tool for checking the coding style?

CaptainVincent replied:
Yes, i will accomplish it later.

hydai commented on src/fund.rs:
```diff
@@ -0,0 +1,63 @@
+    pub fn try_unlock(timestamp:u64, state:&mut State<NullBackend>) {
+        let beneficiary = Address::from_str(FundManager::BENEFICIARY).unwrap();
+        let value = state.storage_at(&beneficiary, &H256::from(0)).unwrap();
+        let init_timestamp = value.get(24..).unwrap().read_i64::<BigEndian>().unwrap();
```
What does the `24` stand for?

CaptainVincent replied:
get byte slice from position 24 to 32 (last 8 bytes). use it and convert to i64.

hydai replied:
We should add a comment to describe the data layout?

hydai commented on src/fund.rs:
```diff
@@ -0,0 +1,63 @@
+        let value = state.storage_at(&beneficiary, &H256::from(0)).unwrap();
+        let init_timestamp = value.get(24..).unwrap().read_i64::<BigEndian>().unwrap();
+        let value = state.storage_at(&beneficiary, &H256::from(1)).unwrap();
+        let mut unlock_round = value.get(28..).unwrap().read_u32::<BigEndian>().unwrap();
```
What does the `28` stand for?

CaptainVincent replied:
get byte slice from position 28 to 32 (last 4 bytes). use it and convert to u32.

hydai replied:
Ditto.

hydai commented on src/fund.rs:
```diff
@@ -0,0 +1,63 @@
+        let value = state.storage_at(&beneficiary, &H256::from(1)).unwrap();
+        let mut unlock_round = value.get(28..).unwrap().read_u32::<BigEndian>().unwrap();
+        let value = state.storage_at(&beneficiary, &H256::from(2)).unwrap();
+        let mut unlock_amount = value.get(28..).unwrap().read_u32::<BigEndian>().unwrap();
```
What does the `28` stand for?

CaptainVincent replied:
same as before comment.

hydai replied:
Ditto.

hydai commented on src/fund.rs:
```diff
@@ -0,0 +1,63 @@
+    /// Initial unlock bucket.
+    const INITIAL_BUCKET: f32 = 525000.0;
+
+    pub fn try_unlock(timestamp:u64, state:&mut State<NullBackend>) {
```
Should we make this function atomic?
Once it failed, we should rollback states.

CaptainVincent replied:
I think this case should be covered by oasis-parity state.commit function (I saw oasis-parity use storage_cache internal) and consensus layer.
If some compute node fail within execution then its' state change should not be acceptable by others and need drop & re-sync.
Do my opinion is reasonable?

hydai replied:
It makes sense.

CaptainVincent commented on src/fund.rs:
```diff
@@ -0,0 +1,63 @@
+            last_ym.add_months(unlock_round);
+            let cur_ym = timestamp_to_mt(timestamp as i64);
+            let mut funding = 0;
+            while last_ym <= cur_ym && unlock_amount < FundManager::TOTAL_AMOUNT{
```
I think I should use `last_ym < cur_ym` here. I will fixup it.

hydai commented on src/fund.rs:
```diff
@@ -0,0 +1,90 @@
+        let value = state.storage_at(&beneficiary, &H256::from(2)).unwrap();
+        let mut unlock_amount = value.get(28..).unwrap().read_u32::<BigEndian>().unwrap();
+
+        if unlock_amount < FundManager::TOTAL_AMOUNT {
```
We should also add comments to explain our unlock logic.

dm4 commented on src/fund.rs:
```diff
@@ -0,0 +1,94 @@
+    // Factor
+    const FACTOR: f32 = 0.5;
+    /// Total Amount.
+    const TOTAL_AMOUNT: u32 = 21000000;
```
Should we use `wei` as unit here?

dm4 replied:
After using `wei` as unit, we should use `u256` here.

CaptainVincent replied:
done.

dm4 commented on src/fund.rs:
```diff
@@ -0,0 +1,94 @@
+    /// Total Amount.
+    const TOTAL_AMOUNT: u32 = 21000000;
+    /// Initial unlock bucket.
+    const INITIAL_BUCKET: f32 = 525000.0;
```
Should we use `wei` as unit and use `u256` as type here?

CaptainVincent committed:
[Runtime] Change amount unit from eth to wei

hydai commented on src/fund.rs:
```diff
@@ -0,0 +1,93 @@
+use date_time::month_tuple::MonthTuple;
+use std::{convert::TryInto, str::FromStr};
+
+fn timestamp_to_mt(timestamp: i64) -> MonthTuple {
```
We need more comments to explain this function.

CaptainVincent replied:
done.

hydai commented on src/fund.rs:
```diff
@@ -0,0 +1,93 @@
+            let mut funding = U256::from(0);
+            let mut exponent = 0;
+            let mut bucket = initial_bucket;
+            while last_ym < cur_ym && unlock_amount < total_amount {
```
I believe there should be comments to explain the details of this while loop.

CaptainVincent replied:
done.

hydai commented on src/fund.rs:
```diff
@@ -0,0 +1,93 @@
+        let mut unlock_amount: U256 = value.into();
+
+        if unlock_amount < total_amount {
+            let mut last_ym = timestamp_to_mt(init_timestamp as i64);
```
You variable name should have more comments.

CaptainVincent replied:
done.

hydai reviewed (dismissed):
Please add more comments. Thanks.

dm4 commented on src/fund.rs:
```diff
@@ -0,0 +1,100 @@
+
+    // There is a total fixed supply of 21 million OETHs.
+    // The blockchain unlocks 525,000 OETHs every month in the first 20 months and
+    // the monthly release is cut to 1/2 every 20 months.
```
Should we move these comment before variable declaration?

CaptainVincent replied:
done.

dm4 commented on src/fund.rs:
```diff
@@ -0,0 +1,100 @@
+    const BENEFICIARY: &'static str = "7110316b618d20d0c44728ac2a3d683536ea682b";
+    /// Adjust period.
+    const PERIOD: u32 = 20;
+    // Factor
```
Why only factor use `//` comment? XD

CaptainVincent replied:
my mistake.

CaptainVincent replied:
done.

dm4 commented on src/fund.rs:
```diff
@@ -0,0 +1,100 @@
+                unlock_round += 1;
+                if exponent != unlock_round / FundManager::PERIOD {
+                    exponent = unlock_round / FundManager::PERIOD;
+                    bucket = initial_bucket / U256::from(FundManager::FACTOR.pow(exponent));
```
`FACTOR.pow(exponent)` will overflow if `exponent > 31`

CaptainVincent replied:
Yes, you are right.
Do we need consider this scenario?
This case means (20 months)*31 / 12 months almost equal 52 years.
If we try to cover all cases, at least need 10 bytes integer.
(525000000000000000000000 wei = 0x6F2C4E995EC98E200000)

hydai replied:
We should avoid any overflow issue even if it may happen in the future.

CaptainVincent replied:
done. I use type U256 handle it finally.

dm4 commented on src/fund.rs:
```diff
@@ -0,0 +1,100 @@
+            // still have funding not yet unlocked.
+            while last_ym < cur_ym && unlock_amount < total_amount {
+                last_ym.add_months(1);
+                unlock_round += 1;
```
If we update `unlock_round` here, we'll miss 1 round and unlock 525,000 OETH only for 19 rounds.

CaptainVincent replied:
done.

hydai commented:
@CaptainVincent Could you add a test to cover this feature? Using `#[test]`, so we can run tests with `cargo test`.

CaptainVincent committed:
[Test] Add test code

1. Add test cases and use `cargo test --features="test"` cmd to testing.
2. Remove needless storage variable.

CaptainVincent commented:
> @CaptainVincent Could you add a test to cover this feature? Using `#[test]`, so we can run tests with `cargo test`.

done.

hydai reviewed (approved):
LGTM

CaptainVincent committed:
[Runtime] Minor update unlock token strategies

1. Use fixed 30 days as unit round. The release amount will cut to 1/2 every 20 rounds.
2. Release amount in a single round is depending on time range. (linear)

dm4 reviewed (approved):
Nice pic! LGTM!

hydai reviewed (approved):
Let's launch the testnet for unlocking mechanism next week. Well done! 👍
